### PR TITLE
Chunk status alert

### PR DESCRIPTION
**Pull request by zomabies**
This PR add some feature for chunk status, as follow.

Indicate if current chunk is unloaded, use last ground height as reference.

Player will receive an alert when they fly into unloaded chunk after a certain amount of time. After 3.2 seconds, alert will be shown and will pitch up by ~10 degrees if enabled.

Screenshot
image

### CODE REVIEW

File: src/main/java/ru/octol1ttle/flightassistant/FAModMenuImpl.java [hunk 1]
```diff
@@ -330,6 +330,18 @@ private ConfigCategory computers(Text name, ComputerConfig config, ComputerConfi
                         .controller(TickBoxControllerBuilder::create)
                         .build())
 
+                .option(LabelOption.create(Text.translatable("config.flightassistant.computers.chunk_state")))
+                .option(Option.<ComputerConfig.WarningMode>createBuilder()
+                        .name(Text.translatable("config.flightassistant.computers.chunk_state.warning"))
+                        .binding(defaults.chunkStatusWarning, () -> config.chunkStatusWarning, o -> config.chunkStatusWarning = o)
+                        .controller(opt -> EnumControllerBuilder.create(opt).enumClass(ComputerConfig.WarningMode.class))
+                        .build())
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/FAModMenuImpl.java]: Normal warnings don't need a warning mode setting

File: src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import ru.octol1ttle.flightassistant.alerts.BaseAlert;
+import ru.octol1ttle.flightassistant.alerts.IECAMAlert;
+import ru.octol1ttle.flightassistant.computers.AirDataComputer;
+import ru.octol1ttle.flightassistant.computers.safety.ChunkStatusComputer;
+import ru.octol1ttle.flightassistant.computers.safety.ChunkStatusComputer.ChunkStatusLevel;
+import ru.octol1ttle.flightassistant.config.FAConfig;
+
+public class UnloadedChunkAlert extends BaseAlert implements IECAMAlert {
+
+    private final AirDataComputer adc;
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java]: Should be `data`

File: src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        final int textWidth = textRenderer.getWidth(alertText) + 5;
+        final String seconds = "%.1f".formatted(chunkStatus.getLastDiffMS() * 0.001f);
+        final Text text = Text.translatable("alerts.flightassistant.unloaded_chunk_time", seconds);
+        HudComponent.drawHighlightedText(textRenderer, context, text, x + textWidth, y, chunkStatus.getIndicator(ChunkStatusLevel.WARNING), highlight);
+        return 1;
+    }
+
+    @Override
+    public @NotNull AlertSoundData getSoundData() {
+        return AlertSoundData.ifEnabled(FAConfig.computer().chunkStatusWarning, AlertSoundData.MASTER_CAUTION);
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java]: Alerts in red shouldn't be combined with the MASTER CAUTION sound (and CAUT doesn't seem appropriate for how dangerous flying in unloaded chunks can be)

File: src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            return 0;
+        }
+
+        final MutableText alertText = Text.translatable("alerts.flightassistant.unloaded_chunk");
+        HudComponent.drawText(textRenderer, context, alertText, x, y, chunkStatus.getIndicator(ChunkStatusLevel.CAUTION));
+
+        final int textWidth = textRenderer.getWidth(alertText) + 5;
+        final String seconds = "%.1f".formatted(chunkStatus.getLastDiffMS() * 0.001f);
+        final Text text = Text.translatable("alerts.flightassistant.unloaded_chunk_time", seconds);
+        HudComponent.drawHighlightedText(textRenderer, context, text, x + textWidth, y, chunkStatus.getIndicator(ChunkStatusLevel.WARNING), highlight);
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java]: The timer doesn't appear important to the user

File: src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java [hunk 5]
```diff
@@ -48,6 +51,10 @@ public void tick() {
         } else {
             smoothSetPitch(targetPitch, time.deltaTime);
         }
+
+        if (chunkStatus.shouldCorrectTerrain()){
+            smoothSetPitch(FAMathHelper.toDegrees(MathHelper.atan2(chunkStatus.recoverVector.y, chunkStatus.recoverVector.x)), time.deltaTime);
+        }
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java]: This instance of automatic pitch correction isn't mutually exclusive with autopilot pitch movement, which will lead to these systems fighting

File: src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java [hunk 6]
```diff
@@ -48,6 +51,10 @@ public void tick() {
         } else {
             smoothSetPitch(targetPitch, time.deltaTime);
         }
+
+        if (chunkStatus.shouldCorrectTerrain()){
+            smoothSetPitch(FAMathHelper.toDegrees(MathHelper.atan2(chunkStatus.recoverVector.y, chunkStatus.recoverVector.x)), time.deltaTime);
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java]: Why calculate pitch from a vector if the vector is constant?

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java:1]: This computer seems overengineered, there's instances of data duplication across computers and unnecessary logic to compute the unloaded time which, as outlined previously, doesn't bring any value

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                return ChunkStatusLevel.ADVISORY;
+            }
+        }
+        return ChunkStatusLevel.NORMAL;
+    }
+
+
+    public Color getIndicator(ChunkStatusLevel maxLevel) {
+        return switch (currentStatus.getHighestLevel(maxLevel)) {
+            case ADVISORY -> FAConfig.indicator().advisoryColor;
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: The advisory color is to show things that the user needs to follow. Examples are ECAM actions, flight directors and target altitude/heading. It is not to be used for warnings

File: src/main/java/ru/octol1ttle/flightassistant/indicators/AltitudeIndicator.java [hunk 8]
```diff
@@ -54,7 +54,12 @@ public void render(DrawContext context, TextRenderer textRenderer) {
 
         if (FAConfig.indicator().showGroundAltitude) {
             Color color = data.altitude() < safeLevel ? FAConfig.indicator().warningColor : FAConfig.indicator().frameColor;
-            drawText(textRenderer, context, Text.translatable(data.groundLevel == data.voidLevel() ? "flightassistant.void_level" : "flightassistant.ground_level"), xAltText - 10, bottom, color);
+            if (!data.isCurrentChunkLoaded) {
+                drawText(textRenderer, context, Text.literal("?"), xAltText - 10, bottom, FAConfig.indicator().cautionColor);
+            }
+            else {
+                drawText(textRenderer, context, Text.translatable(data.groundLevel == data.voidLevel() ? "flightassistant.void_level" : "flightassistant.ground_level"), xAltText - 10, bottom, color);
+            }
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/indicators/AltitudeIndicator.java]: This should show a placeholder instead. similar to those displayed when an indicator failure occurs (see `renderFaulted`)

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java [hunk 9]
```diff
@@ -45,9 +48,12 @@ public void tick() {
             smoothSetPitch(90.0f, MathHelper.clamp(time.deltaTime / gpws.descentImpactTime, 0.001f, 1.0f));
         } else if (gpws.shouldCorrectTerrain()) {
             smoothSetPitch(FAMathHelper.toDegrees(MathHelper.atan2(gpws.terrainAvoidVector.y, gpws.terrainAvoidVector.x)), time.deltaTime);
+        } else if (chunkStatus.shouldCorrectTerrain()) {
+            smoothSetPitch(chunkStatus.recoverPitch, time.deltaTime);
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java]: Make `VoidLevelComputer.OPTIMUM_ALTITUDE_PRESERVATION_PITCH` public and use it here

File: src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import net.minecraft.text.Text;
+import org.jetbrains.annotations.NotNull;
+import ru.octol1ttle.flightassistant.HudComponent;
+import ru.octol1ttle.flightassistant.alerts.AlertSoundData;
+import ru.octol1ttle.flightassistant.alerts.BaseAlert;
+import ru.octol1ttle.flightassistant.alerts.IECAMAlert;
+import ru.octol1ttle.flightassistant.computers.safety.ChunkStatusComputer;
+
+public class UnloadedChunkAlert extends BaseAlert implements IECAMAlert {
+
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java]: Unnecessary newline

File: src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    public int render(TextRenderer textRenderer, DrawContext context, int x, int y, boolean highlight) {
+        return HudComponent.drawHighlightedText(textRenderer, context, Text.translatable("alerts.flightassistant.unloaded_chunk"), x, y,
+                chunkStatus.getIndicator(), highlight);
+    }
+
+    @Override
+    public @NotNull AlertSoundData getSoundData() {
+        return AlertSoundData.MASTER_WARNING;
+    }
+
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/alerts/nav/UnloadedChunkAlert.java]: Unnecessary newline

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import ru.octol1ttle.flightassistant.computers.AirDataComputer;
+import ru.octol1ttle.flightassistant.computers.ComputerHost;
+import ru.octol1ttle.flightassistant.computers.ITickableComputer;
+import ru.octol1ttle.flightassistant.computers.TimeComputer;
+import ru.octol1ttle.flightassistant.config.FAConfig;
+
+import java.awt.Color;
+
+public class ChunkStatusComputer implements ITickableComputer {
+
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: Unnecessary newline

File: src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java [hunk 13]
```diff
@@ -45,9 +48,12 @@ public void tick() {
...
         } else {
             smoothSetPitch(targetPitch, time.deltaTime);
         }
+
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/autoflight/PitchController.java]: Unnecessary newline

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        if (!changesPending) {
+            changesPending = true;
+        }
+
+        isLoaded = data.isCurrentChunkLoaded;
+
+        if (isLoaded) {
+            if (!host.faulted.contains(time) && time.prevMillis != null) {
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: Don't inject `ComputerHost` here and remove the faulted check

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        if (!data.isFlying()) {
+            if (changesPending) {
+                reset();
+            }
+            return;
+        }
+
+        if (!changesPending) {
+            changesPending = true;
+        }
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: How does this differ from 
```java
        if (!data.isFlying()) {
            reset();
            return;
        }
```

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+        isLoaded = data.isCurrentChunkLoaded;
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: Data duplication, callers should use `data.isCurrentChunkLoaded` directly

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public void reset() {
+        isInWarning = false;
+        isLoaded = true;
+        lastDiffMS = 0f;
+        lastEncounteredMS = 0f;
+        offsetMS = 0f;
+        changesPending = false;
+    }
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: Overriden methods should come last

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+        return isInWarning;
+    }
+
+    public float getLastDiffMS() {
+        return lastDiffMS;
+    }
+
+    public Color getIndicator() {
+        return FAConfig.indicator().warningColor;
+    }
```
Comment (Octol1ttle) [inline src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: Callers should use `FAConfig.indicator().warningColor` directly

**Review by Octol1ttle — CHANGES_REQUESTED** [review]

File: src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java (hunk 15 above)
Comment (zomabies) [reply src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: There's no difference, but preferably it should only run once when stop flying.

Comment (Octol1ttle) [reply src/main/java/ru/octol1ttle/flightassistant/computers/safety/ChunkStatusComputer.java]: I think it's fine to always run it for sake of simplicity

**Review by Octol1ttle — APPROVED** [review]

**Pull request merged**